Marlin/src/lcd/extui/lib/mks_ui/wifi_module.cpp

```diff
@@ -1817,7 +1817,7 @@ void get_wifi_commands() {
 
         #if DISABLED(EMERGENCY_PARSER)
           // Process critical commands early
-          switch(HASH(command)) {
+          switch (HASH(command)) {
```

Comment:
Damn it. I swear I had the extension set up correctly.

Reply:
The real question is, how well does this 'constexpr hash' thing work on SRAM string `command`?

Reply:
The constexpr *disappears* when the calling context is not compile time. So it's a basic `a * prime + accumulator` hash function here.

Reply:
Note that most commands (where we care about speed) start with **G** so if computing the hash is slower than comparing a single character, this will not be efficient, especially added up over the life of a print job.

Reply:
And how about hash collisions when using the HASH thing with runtime text (user/host input, for example) ? How often we would pull your hair off with weird and hard to track bugs?

Reply:
See [here](https://godbolt.org/z/3jE137) for the compiled output of such method: TLDR: it's short and optimal.

Reply:
More straightforward for situations like this would be a compile-time function like this…
```cpp
if (small_strcmp(strptr, "M108")) { . . . }
```
That converts into this…
```cpp
if (strptr[0] == 'M' && strptr[1] == '1' && strptr[2] == '0' && strptr[3] == '8' && strptr[4] == '\0') { . . . }
```

Reply:
It's not being used for G code parsing. I'm not sure it would make sense here, the codes are too short to worth it. It's being used where it matters: for matching long strings. (I've written some "when to use it" documentation in the PR).

About hash collisions: I wouldn't trust this to build a dictionary. But here we're checking a small subset of possible case (the cases in a switch is low). To have a collision with the hash function chosen, you'll need a very large string (you need to overflow the 32 bits accumulator 33 times here to collide). For such small subset, you'll hit the string buffer length limit before you succeed colliding the hash value. 

It's not like a hash table where you fold back the hash function results with a modulo into a very limited subset of buckets.

Reply:
Indeed for this emergency parsing, I could have skipped it, it does not make much sense, since the hash value takes 4 bytes and the "Mxxx" code takes 5 bytes.

Reply:
I'll revert this part.

Reply:
You'll need to add the "not" zero test for each char in your example. I'm not sure it's faster to have a bunch of test & branching compared to a streamlined add & shift function.

Reply:
There's one case that might worth it, it's when doing progressive parsing:
```
if (str[0] == 'M') {
  if (str[1] =='1') {
    if (str[2] == '1' && str[3] == '2') {
       // M112
    } else if (str[2] == '2' && str[3] == '3') { // M123 }
  } else if (str[1] == '2' && str[3] == '4') { // M24 }
} etc...

Here you can skip checking for zeros and you only test a char once, but such code is a real PITA to maintain, even if it's optimal. That's what Lex and Bison are for.

Reply:
My name "victor" generated the value `6.953.986.493.308` (using the link you shared here).... 4 letters it overflow the 32bits accumulator. I'm not sure we need 33 times this to generate a colision.

Reply:
The compiler should automatically rearrange a series of individual character comparisons into this result. If my constexpr `str_match` function is constructed correctly, then it should ultimately result in this exact set of comparisons.

Reply:
> More straightforward for situations like this would be a compile-time function like this…
> 
> ```c++
> if (small_strcmp(strptr, "M108")) { . . . }
> ```
> 
> That converts into this…
> 
> ```c++
> if (strptr[0] == 'M' && strptr[1] == '1' && strptr[2] == '0' && strptr[3] == '8' && strptr[4] == '\0') { . . . }
> ```

It will only worth such optimization when used on intensive cpu operations, isn’t?
I mean, we are using about a few cycles less then using a normal strcmp...

Reply:
int i = constHash("Yict");
 int j = constHash("zhct");

both values:
2090753662
2090753662

by hand...

Reply:
> It will only worth such optimization when used on intensive cpu operations, isn’t?
> I mean, we are using about a few cycles less then using a normal strcmp...

You won't outsmart the compiler here. It'll always produce the best code using plain old strcmp.

Reply:
> int i = constHash("Yict");
> int j = constHash("zhct");

You're right, you've broken all my dreams...

The value for the prime number must be higher than the highest code index we use, it should be 257 not 33. I've changed it by 33 to use less value, but it's plain wrong as you spotted.

The equation the hash function must resolve is this: `x2 + primeNumber * x1 = y2 + primeNumber * y1`
To find a collision, it means that you need to find a tuple (x2, x1) that gives the same results as (y2, y1). With a primeNumber set to 33, it's possible, you just need to use a char that's 33 slot higher than the previous (that's what you did). 

With 257, you can't have such value, so you shouldn't be able to find a collision... unless wrapping the accumulator.

Reply:
The [godbolt to prove it](https://godbolt.org/z/8xYTn6)